lib/Common/Memory/IdleDecommitPageAllocator.cpp

```diff
@@ -96,7 +117,26 @@ IdleDecommitPageAllocator::LeaveIdleDecommit(bool allowTimer)
 #ifdef IDLE_DECOMMIT_ENABLED
     if (allowTimer)
     {
-        cs.Enter();
+        if (!cs.TryEnter())
+        {
+            this->tryEnterIdleDecommit = true;
+
+#ifdef ENABLE_BASIC_TELEMETRY
+            decommitStats->lastEnterLeaveIdleDecommitTick = Js::Tick::Now();
+#endif
+
+            cs.Enter();
+
+#ifdef ENABLE_BASIC_TELEMETRY
+            decommitStats->lastEnterLeaveIdleDecommitCSWaitTime = Js::Tick::Now() - decommitStats->lastEnterLeaveIdleDecommitTick;
+            decommitStats->totalEnterLeaveIdleDecommitCSWaitTime += decommitStats->lastEnterLeaveIdleDecommitCSWaitTime;
+            if (decommitStats->lastEnterLeaveIdleDecommitCSWaitTime > decommitStats->maxEnterLeaveIdleDecommitCSWaitTime)
+            {
+                decommitStats->maxEnterLeaveIdleDecommitCSWaitTime = decommitStats->lastEnterLeaveIdleDecommitCSWaitTime;
+            }
+#endif
+            this->tryEnterIdleDecommit = false;
```

Comment:
Is there any concern that resetting this could be skipped (e.g., an exception is raised)?  Would it make more sense to use an "auto-reset object" and restore this in the destructor?

Reply:
Good idea. I have added a auto-reset object.

---
In reply to: [196262873](https://github.com/Microsoft/ChakraCore/pull/5326#discussion_r196262873) [](ancestors = 196262873)